File: euler48.py
# Найдите последние десять цифр суммы 1^1 + 2^2 + 3^3 + ... + 1000^1000.

from datetime import datetime


def euler48() -> None:
    """
    Решение задачи Эйлера №48.

    Найдите последние десять цифр суммы 1^1 + 2^2 + 3^3 + ... + 1000^1000.
    """
    n = 10

    # 1 вариант
    start_time = datetime.now()
    long_number = 0
    for i in range(1, 1001):
        long_number += i ** i
    print(str(long_number)[-n:])
    print(datetime.now() - start_time)

    # 2 вариант
    start_time = datetime.now()
    result, long_number = 0, 10 ** n
    for i in range(1, 1001):
        temp = i
        for j in range(1, i):
            temp *= i
            temp %= long_number
        result += temp
        result %= long_number
    print(result)
    print(datetime.now() - start_time)

    # 3 вариант
    start_time = datetime.now()
    print(str(sum(i ** i for i in range(1, 1001)))[-n:])
    print(datetime.now() - start_time)


if __name__ == "__main__":
    euler48()

File: go/euler48.go
// euler48 находит последние десять цифр суммы 1^1 + 2^2 + ... + 1000^1000.
//
// Решение задачи Эйлера №48.
package main

import (
	"fmt"
	"math/big"
	"time"
)

func euler48() {
	n := 10

	startTime := time.Now()
	longNumber := big.NewInt(0)
	mod := big.NewInt(1)
	for i := 0; i < n; i++ {
		mod.Mul(mod, big.NewInt(10))
	}
	for i := 1; i <= 1000; i++ {
		exp := big.NewInt(int64(i))
		base := big.NewInt(int64(i))
		pow := big.NewInt(0)
		pow.Exp(base, exp, mod)
		longNumber.Add(longNumber, pow)
		longNumber.Mod(longNumber, mod)
	}
	fmt.Println(fmt.Sprintf("%d", longNumber))
	fmt.Println(time.Since(startTime))

	startTime = time.Now()
	result := 0
	longNumber = big.NewInt(0)
	for i := 1; i <= 1000; i++ {
		temp := big.NewInt(int64(i))
		exp := big.NewInt(int64(i))
		temp.Exp(temp, exp, mod)
		result += int(temp.Int64())
		result %= 10000000000
	}
	fmt.Println(result)
	fmt.Println(time.Since(startTime))

	startTime = time.Now()
	sum := big.NewInt(0)
	for i := 1; i <= 1000; i++ {
		exp := big.NewInt(int64(i))
		base := big.NewInt(int64(i))
		pow := big.NewInt(0)
		pow.Exp(base, exp, nil)
		sum.Add(sum, pow)
	}
	str := fmt.Sprintf("%d", sum)
	if len(str) >= n {
		fmt.Println(str[len(str)-n:])
	} else {
		fmt.Println(str)
	}
	fmt.Println(time.Since(startTime))
}

func main() {
	euler48()
}
